accumulate.go

```diff
@@ -31,66 +40,121 @@ namespace are stored.
+			continue
+		}
+
+		err = accumulateBucketMetricsForObjectsUser(accumulated, bucketMetricsData, organization, ns)
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s: %v\n", bucketMetricsData.Subject.BucketName, err)
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s: %v\n", name, err)
 			continue
 		}
 	}
 
 	return accumulated, nil
 }
 
-func accumulateBucketMetricsForObjectsUser(accumulated map[AccumulateKey]uint64, bucketMetricsData cloudscale.BucketMetricsData, objectsUser *cloudscale.ObjectsUser) error {
-	if len(bucketMetricsData.TimeSeries) != 1 {
-		return fmt.Errorf("There must be exactly one metrics data point, found %d", len(bucketMetricsData.TimeSeries))
+func fetchBuckets(ctx context.Context, k8sclient client.Client, opts ...client.ListOption) (map[string]string, error) {
+	buckets := &cloudscalev1.BucketList{}
+	if err := k8sclient.List(ctx, buckets, opts...); err != nil {
+		return nil, fmt.Errorf("bucket list: %w", err)
+	}
+
+	bucketNS := map[string]string{}
+	for _, b := range buckets.Items {
+		bucketNS[b.Spec.ForProvider.BucketName] = b.Labels[namespaceLabel]
+	}
+
+	if buckets.Continue != "" {
+		m, err := fetchBuckets(ctx, k8sclient, client.Continue(buckets.Continue))
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range m {
+			bucketNS[k] = v
+		}
+	}
+	return bucketNS, nil
+}
+
+func fetchNamespaces(ctx context.Context, k8sclient client.Client, opts ...client.ListOption) (map[string]string, error) {
+	namespaces := &corev1.NamespaceList{}
+	if err := k8sclient.List(ctx, namespaces, opts...); err != nil {
+		return nil, fmt.Errorf("namespace list: %w", err)
 	}
 
-	tenantStr := objectsUser.Tags["tenant"]
-	if tenantStr == "" {
-		return fmt.Errorf("no tenant information found on objectsUser")
+	nsOrganizations := map[string]string{}
+	for _, ns := range namespaces.Items {
+		nsOrganizations[ns.Name] = ns.Labels[organizationLabel]
 	}
-	namespace := objectsUser.Tags["namespace"]
-	if namespace == "" {
-		return fmt.Errorf("no namespace information found on objectsUser")
+
+	if namespaces.Continue != "" {
+		m, err := fetchNamespaces(ctx, k8sclient, client.Continue(namespaces.Continue))
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range m {
+			nsOrganizations[k] = v
+		}
 	}
+	return nsOrganizations, nil
+}
+
+func accumulateBucketMetricsForObjectsUser(accumulated map[AccumulateKey]uint64, bucketMetricsData cloudscale.BucketMetricsData, organization, namespace string) error {
+	if len(bucketMetricsData.TimeSeries) != 1 {
+		return fmt.Errorf("there must be exactly one metrics data point, found %d", len(bucketMetricsData.TimeSeries))
+	}
+
 	// For now all the buckets have the same zone. This may change in the future if Cloudscale decides to have different
 	// prices for different locations.
 	zone := sourceZones[0]
 
 	sourceStorage := AccumulateKey{
 		Query:     sourceQueryStorage,
 		Zone:      zone,
-		Tenant:    tenantStr,
+		Tenant:    organization,
```

Comment:
now that I see this: should I keep calling that `tenant` or is `organization` better and I should switch all to that?

Reply:
yes, I'd say keep tenant, as that is the term used for and in the billing system.
see also https://kb.vshn.ch/appuio-cloud/references/architecture/metering-data-flow.html